keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,534 @@
+doesn't change the current behavior.
+Note that step 2 doesn't pay attention to deletion timestamps. So the Job
+controller might account the Pod in the lists and counters if the Pods
+enter Failed or Succeeded phase before or while being deleted.
+<<[/UNRESOLVED]>>
+   
+### Pod adoption
+
+If a Job with `.status.uncountedTerminatedPodUIDs.enabled = true` can adopt a Pod
+(according to the existing adoption criteria), this Pod might not have a
+finalizer.
+
+The job controller adds the finalizer in the same patch request that modifies
+the owner reference.
+
+### Test Plan
+
+- Unit tests:
+  - Job sync with feature gate enabled.
+  - Removal of finalizers when feature gate is disabled.
+  - Tracking of terminating Pods.
+- Integration tests:
+  - Job tracking with feature enabled.
+  - Tracking of terminating Pods.
+  - Transition from feature enabled to disabled and enabled again.
+- E2E test:
+  - Job tracking with feature enabled.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implementation:
+  - Job tracking without lingering Pods
+  - Removal of finalizer when feature gate is disabled.
+- Tests: unit, integration, E2E
+
+#### Alpha -> Beta Graduation
+
+- Implementation:
+  - Index tracking for [Indexed Jobs](http://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job)
+- Metrics:
+  - latency
+  - errors
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- E2E test graduates to conformance.
+- Job tracking scales to 10^5 completions.
+
+### Upgrade / Downgrade Strategy
+
+When the feature `JobTrackingWithoutLingeringPods` is enabled for the first
+time, the cluster can have Jobs whose Pods don't have the
+`batch.kubernetes.io/job-completion` finalizer. It would be hard to add the
+finalizer to all Pods while preventing race conditions.
+
+We use `.status.uncountedTerminatedPodUIDs.enabled` to indicate whether the Job
+was created after the feature was enabled. If this boolean is false, the Job
+controller tracks Pods using the legacy algorithm.
+
+The kube-apiserver sets `.status.uncountedTerminatedPodUIDs.enabled = true` when
+the feature gate `JobTrackingWithoutLingeringPods` is enabled, at Job creation.
+
+When the feature is disabled after being enabled for some time, the next time
+the Job controller syncs a Job:
+1. It removes finalizers from all Pods owned by the Job.
+2. Sets `.status.uncountedTerminatedPodUIDs.enabled = false`
+
+### Version Skew Strategy
+
+No implications to node runtime.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: JobTrackingWithoutLingeringPods
+    - Components depending on the feature gate:
+      - kube-apiserver
+      - kube-controller-manager
```

Comment:
I'd probably leave it only on kcm side, no need to touch kas.

Reply:
we do need to set the struct to non-nil when the feature gate is enabled.

Reply:
Nope, it's not needed, we don't do it for any of the controllers. It's controller's responsibility to set status accordingly.

Reply:
Leaving as discussed offline